File: 128-longest-consecutive-sequence/128-longest-consecutive-sequence.scala
import scala.collection.mutable._

object Solution {
    def longestConsecutive(nums: Array[Int]): Int = {
        val set = nums.toSet
        val map: Map[Int, Int] = Map()

        var res = 0

        set.foreach { n =>
            var l = 1
            var c = n

            while(!map.contains(c + 1) && set.contains(c + 1)) {
                l += 1
                c += 1
            }

            l += map.get(c + 1).getOrElse(0)

            map += (n -> l)

            res = res max l
        }

        res
    }
}

File: 128-longest-consecutive-sequence/128-longest-consecutive-sequence.java
class Solution {
    public int longestConsecutive(int[] nums) {
        int res = 0;
        
        Set<Integer> set = new HashSet<>();
        
        for(int n : nums) {
            set.add(n);
        }
        
        Map<Integer, Integer> map = new HashMap<>();
        
        for(int n : nums) {
            if (map.containsKey(n)) continue;
            
            int cur = n, con = 1;
            while(set.contains(cur + 1)) {
                if (map.containsKey(cur + 1)) {
                    con += map.get(cur + 1);
                    break;
                }
                con++;
                cur++;
            }
            
            map.put(n, con);
            
            res = Math.max(res, con);
        }
        
        return res;
    }
}